### PR TITLE
Implement realm server /_user endpoint for credits balance

### PR DESCRIPTION
**Pull request by FadhlanR**
In this PR, I implemented a new endpoint to retrieve the user's credit balance and integrated it into the host application. 

https://github.com/user-attachments/assets/1058d0d7-73a3-4bfb-b9e4-aaa5b5e334f0

https://github.com/user-attachments/assets/5eb94812-b8ce-450a-8a67-d30bf6986042

### CODE REVIEW

File: packages/host/app/components/operator-mode/profile-info-popover.gts [hunk 1]
```diff
@@ -192,29 +208,32 @@ export default class ProfileInfoPopover extends Component<ProfileInfoPopoverSign
...
+  private get isLoading() {
+    return this.fetchCreditInfo.isLoading;
+  }
+
   private get plan() {
     return this.fetchCreditInfo.value?.plan;
   }
 
+  private get planName() {
+    return this.plan?.name ?? '-';
```
Comment (jurgenwerk) [inline packages/host/app/components/operator-mode/profile-info-popover.gts]: I think we don't want a case where user can work inside the app and is not on a plan

File: packages/realm-server/tests/realm-server-test.ts [hunk 2]
```diff
@@ -154,7 +163,7 @@ module('Realm Server', function (hooks) {
           copySync(join(__dirname, 'cards'), testRealmDir);
         }
         let virtualNetwork = createVirtualNetwork();
-
+        dbAdapter = _dbAdapter;
```
Comment (backspace) [inline packages/realm-server/tests/realm-server-test.ts]: I’m not up on these tests, this assignment also exists on line 157, is it needed again? If so, maybe an explanation comment would help.

File: packages/realm-server/server.ts [hunk 3]
```diff
@@ -665,6 +727,122 @@ export class RealmServer {
...
+    } as FetchUserResponse;
+
+    return setContextResponse(
+      ctxt,
+      new Response(JSON.stringify(responseBody), {
+        headers: { 'content-type': SupportedMimeType.JSONAPI },
+      }),
+    );
+  };
+
```
Comment (backspace) [inline packages/realm-server/server.ts]: Maybe it’s time to break this file up somewhat, the collection of disparate route handlers invites separate files.

File: packages/host/app/services/realm-server.ts [hunk 4]
```diff
@@ -177,25 +178,37 @@ export default class RealmServerService extends Service {
...
+    await this.login();
+    if (this.auth.type !== 'logged-in') {
+      throw new Error('Could not login to realm server');
+    }
+
+    let response = await this.network.authedFetch(`${this.url.origin}/_user`, {
+      headers: {
+        Accept: SupportedMimeType.JSONAPI,
+        'Content-Type': 'application/json',
+        Authorization: `Bearer ${this.token}`,
```
Comment (backspace) [inline packages/host/app/services/realm-server.ts:189]: I’m surprised to see this header as I’d assume `authedFetch` would include something similar, looking at the code it’s not clear to me that it does though, so I’m just confused!

**Review by backspace — APPROVED** [review]

File: packages/realm-server/tests/realm-server-test.ts (hunk 2 above)
Comment (habdelra) [reply packages/realm-server/tests/realm-server-test.ts]: I think it’s an upstream merge artifact.

File: packages/host/app/components/operator-mode/profile-info-popover.gts (hunk 1 above)
Comment (FadhlanR) [reply packages/host/app/components/operator-mode/profile-info-popover.gts]: Should I hide the credit section in account popover when user does not subscribe to any plan?

File: packages/realm-server/tests/realm-server-test.ts (hunk 2 above)
Comment (FadhlanR) [reply packages/realm-server/tests/realm-server-test.ts]: Ah nice catch! I didn't resolve the merge conflict correctly.

File: packages/host/app/components/operator-mode/profile-info-popover.gts (hunk 1 above)
Comment (jurgenwerk) [reply packages/host/app/components/operator-mode/profile-info-popover.gts]: No, I mean that I think user should be subscribed to one of the plans in order to use the app. CS-7396 will ensure user starts with a free plan by subscribing in Stripe, and then they can switch between plans if they want to. I don't see a case here where user is not on one of the available plans.

File: packages/host/app/services/realm-server.ts (hunk 4 above)
Comment (FadhlanR) [reply packages/host/app/services/realm-server.ts:189]: There are two types of authentication: realm authentication and realm server authentication. The `authedFetch` function is used to handle realm authentication. I think using `fetch` here would make it less confusing.

File: packages/host/app/components/operator-mode/profile-info-popover.gts (hunk 1 above)
Comment (jurgenwerk) [reply packages/host/app/components/operator-mode/profile-info-popover.gts]: I guess the question is what will be shown to the existing users that are already registered and didn't go through the stripe flow to subscribe to the free plan. Do we redirect them to the screen in the mentioned ticket, or maybe add a new button in the popover that takes them to the free plan subscribe payment link, e.g. "Subscribe to our free plan to try using Boxel"? It's something to consider and discuss. It's also confusing for user to see they are not on a plan (`-`) and then there is a button to "Upgrade plan" (upgrading from what?).

Comment (FadhlanR) [reply packages/host/app/components/operator-mode/profile-info-popover.gts]: Does that mean in the endpoint we can throw NotFound error if user has no subscription, instead of returning `{ data: null, included: null }` ?

Comment (jurgenwerk) [reply packages/host/app/components/operator-mode/profile-info-popover.gts]: I think it's ok if you return the user data (matrixUserId,..) without subscription data included (or empty subscriptions array)

File: packages/host/app/services/realm-server.ts (hunk 4 above)
Comment (FadhlanR) [reply packages/host/app/services/realm-server.ts:189]: Updated [c25235d](https://github.com/cardstack/boxel/pull/1756/commits/c25235d6563553f81801950a1cbca87a2dec55a9).

File: packages/realm-server/server.ts (hunk 3 above)
Comment (FadhlanR) [reply packages/realm-server/server.ts]: Updated [7dea872](https://github.com/cardstack/boxel/pull/1756/commits/7dea872772f364d2806a1041aa635cd813896123).

File: packages/host/app/components/operator-mode/profile-info-popover.gts (hunk 1 above)
Comment (FadhlanR) [reply packages/host/app/components/operator-mode/profile-info-popover.gts]: > I guess the question is what will be shown to the existing users that are already registered and didn't go through the stripe flow to subscribe to the free plan. Do we redirect them to the screen in the mentioned ticket, or maybe add a new button in the popover that takes them to the free plan subscribe payment link, e.g. "Subscribe to our free plan to try using Boxel"? It's something to consider and discuss. It's also confusing for user to see they are not on a plan (`-`) and then there is a button to "Upgrade plan" (upgrading from what?).


@lukemelia  we’d appreciate your input.

Comment (lukemelia) [reply packages/host/app/components/operator-mode/profile-info-popover.gts]: A user should not be able to access Boxel operator mode unless they have a plan. Users who do not should be shown a message explaining this with a button to sign up for the free plan.

File: packages/realm-server/tests/index.ts [hunk 5]
```diff
@@ -1,14 +1,14 @@
...
+// import './indexing-test';
+// import './loader-test';
+// import './module-syntax-test';
+// import './permissions/permission-checker-test';
+// import './queue-test';
 import './realm-server-test';
-import './virtual-network-test';
-import './billing-test';
+// import './virtual-network-test';
+// import './billing-test';
```
Comment (backspace) [inline packages/realm-server/tests/index.ts]: This was probably meant to be temporary?

**Review by backspace — DISMISSED** [review]
The file splitting looks good, half the tests are missing though 😯

File: packages/realm-server/tests/index.ts (hunk 5 above)
Comment (FadhlanR) [reply packages/realm-server/tests/index.ts]: Oops 😓

**Comment by backspace** [comment]
I merged in from `main` which got rid of the Percy diffs but there’s a [failing realm server test](https://github.com/cardstack/boxel/actions/runs/11728164677/job/32673865641?pr=1756#step:11:2062) that does seem likely related to the code in this PR 😯

File: packages/host/app/components/operator-mode/profile-info-popover.gts (hunk 1 above)
Comment (FadhlanR) [reply packages/host/app/components/operator-mode/profile-info-popover.gts]: @jurgenwerk, I've updated the code so that the account popover no longer handles the 'no plan' case. I also added a note to this ticket: https://linear.app/cardstack/issue/CS-7396/send-user-to-stripe-checkout-for-free-plan-during-registration-flow#comment-aaba938d.

**Comment by FadhlanR** [comment]
> I merged in from `main` which got rid of the Percy diffs but there’s a [failing realm server test](https://github.com/cardstack/boxel/actions/runs/11728164677/job/32673865641?pr=1756#step:11:2062) that does seem likely related to the code in this PR 😯

Fixed

File: packages/host/app/components/operator-mode/profile-info-popover.gts [hunk 6]
```diff
@@ -192,43 +208,50 @@ export default class ProfileInfoPopover extends Component<ProfileInfoPopoverSign
...
+        this.extraCreditsAvailableInBalance == 0)
     );
   }
 
   private get isOutOfPlanCreditAllowance() {
     return (
-      this.creditsUsedInPlanAllowance &&
-      this.creditsIncludedInPlanAllowance &&
-      this.creditsUsedInPlanAllowance >= this.creditsIncludedInPlanAllowance
+      this.creditsAvailableInPlanAllowance == null ||
```
Comment (jurgenwerk) [inline packages/host/app/components/operator-mode/profile-info-popover.gts:248]: Why checking for null? Isn't `this.creditsAvailableInPlanAllowance <= 0` enough to check whether user has any plan credits left?

Comment (FadhlanR) [reply packages/host/app/components/operator-mode/profile-info-popover.gts:248]: Actually, it's just to satisfy the type check, since we decided to not throw NotFoundError from endpoint if no subscription found in database, instead we make creditsAvailableInPlanAllowance value can be null.

File: packages/realm-server/tests/realm-server-test.ts [hunk 7]
```diff
@@ -2587,6 +2596,185 @@ module('Realm Server', function (hooks) {
...
+                {
+                  price: { product: 'prod_creator' },
+                },
+              ],
+            },
+          },
+        },
+      } as StripeInvoicePaymentSucceededWebhookEvent;
+
+      await handlePaymentSucceeded(
```
Comment (jurgenwerk) [inline packages/realm-server/tests/realm-server-test.ts]: Why is this needed? You already set up your user and subscription state needed for this test. I think you shouldn't be setting up state for your test using stripe handlers in this type of test.

Comment (FadhlanR) [reply packages/realm-server/tests/realm-server-test.ts]: You're right, what I need is adding credits for `plan_allowance`. I updated it.

File: packages/realm-server/handlers/handle-fetch-user.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+export default function handleFetchUserRequest({
+  dbAdapter,
+}: CreateRoutesArgs): (ctxt: Koa.Context, next: Koa.Next) => Promise<void> {
+  return async function (ctxt: Koa.Context, _next: Koa.Next) {
+    let token = ctxt.state.token as RealmServerTokenClaim;
+    if (!token) {
+      await sendResponseForSystemError(
+        ctxt,
+        'token is required to create realm',
```
Comment (jurgenwerk) [inline packages/realm-server/handlers/handle-fetch-user.ts]: Message doesn't seem right

File: packages/realm-server/handlers/handle-fetch-user.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+      );
+      return;
+    }
+
+    let { user: matrixUserId } = token;
+    let user = await getUserByMatrixUserId(dbAdapter, matrixUserId);
+    let mostRecentSubscription: Subscription | null = null;
+    if (user) {
```
Comment (jurgenwerk) [inline packages/realm-server/handlers/handle-fetch-user.ts]: Maybe return a 404 here? This way you won't have to have user conditions in your responses

File: packages/realm-server/handlers/handle-fetch-user.ts [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+              attributes: {
+                startedAt: mostRecentSubscription.startedAt,
+                endedAt: mostRecentSubscription.endedAt ?? null,
+                status: mostRecentSubscription.status,
+              },
+              relationships: {
+                plan: {
+                  data: {
+                    type: 'plan',
+                    id: plan!.id,
```
Comment (jurgenwerk) [inline packages/realm-server/handlers/handle-fetch-user.ts]: You use `plan!.` in multiple areas in the response, perhaps it's better to check if the plan exists before you make the response to avoid `!`

File: packages/host/app/components/operator-mode/profile-info-popover.gts [hunk 11]
```diff
@@ -192,43 +208,50 @@ export default class ProfileInfoPopover extends Component<ProfileInfoPopoverSign
...
     return this.fetchCreditInfo.value?.plan;
   }
 
+  private get planName() {
+    return this.plan?.name;
+  }
+
   private get creditsIncludedInPlanAllowance() {
-    return this.fetchCreditInfo.value?.creditsIncludedInPlanAllowance;
+    return this.fetchCreditInfo.value?.plan?.creditsIncluded;
```
Comment (jurgenwerk) [inline packages/host/app/components/operator-mode/profile-info-popover.gts]: I think it's better to show a number here which is a result of `plan_allowance` sum in the ledger for the current billing cycle. The actual allowance could differ from `plan.creditsIncluded` in case when users upgrade plans in the middle of the billing period and credit proration happens. More info here: https://github.com/cardstack/boxel/pull/1776/files#diff-852eaad006d9ce6dbc2b37b0c9835ba58d0e92daf5eebf8bc23a8d4d7428a2d2R330

**Review by jurgenwerk — APPROVED** [review]
handleFetchUserRequest is making 7 requests to the DB. This is a lot and this could be a single query. Given this query will be performed only when user requests it manually (by opening the popover) it's probably ok for now but we might want to optimize it later.
